llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp

```diff
@@ -675,14 +675,14 @@ static bool foldLoadsRecursive(Value *V, LoadOps &LOps, const DataLayout &DL,
       Load2Ptr->stripAndAccumulateConstantOffsets(DL, Offset2,
                                                   /* AllowNonInbounds */ true);
 
-  // Verify if both loads have same base pointers and load sizes are same.
+  // Verify if both loads have same base pointers
   uint64_t LoadSize1 = LI1->getType()->getPrimitiveSizeInBits();
   uint64_t LoadSize2 = LI2->getType()->getPrimitiveSizeInBits();
-  if (Load1Ptr != Load2Ptr || LoadSize1 != LoadSize2)
+  if (Load1Ptr != Load2Ptr)
     return false;
 
-  // Support Loadsizes greater or equal to 8bits and only power of 2.
-  if (LoadSize1 < 8 || !isPowerOf2_64(LoadSize1))
+  // Support Loadsizes greater or equal to 8bits.
+  if (LoadSize1 < 8 || LoadSize2 < 8)
```

Comment:
I think what you want to check at this point is that typeSizeEqualsStoreSize for both loads. Please make sure that something like `i9` plus `i23` does not fold.

The sum being a power of two is already covered by the isTypeLegal check (well, or the target has non-pow2 legal integers, in which case there's no reason not to do the fold).